### PR TITLE
Calculating the AC only component of the samples

### PR DESCRIPTION
**Pull request by synco**
# What does this implement/fix?
Eliminates the need for offset calibration, by using formula for standard-deviation.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#<esphome-docs PR number goes here>

## Test Environment

- [ ] ESP32
- [x] ESP8266

## Example entry for `config.yaml`:
Same

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).
  
If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Review by OttoWinter — COMMENTED** [review]
So, effectively, this approach recalculates the DC offset every measurement interval, where before it was once at setup time and with small adjustments during runtime.

That's ok. My fear is that this will make it less accurate though - at 50Hz with 200ms measurement time we're already just seeing 10 cycles. Any change in amplitude during that time will change the calculated offset, even though we know it's constant. The previous approach was better here because the calculated offset is relatively independent of short changes.

Can we keep the exponential average adjusting of the DC offset? So keep the offset_ field, and adjust it each measurement interval with the calculated `dc` offset.

Another thing that _may_ be an issue is how this handles large float values (for example when measuring in millivolts). When squaring the values get significantly higher (or smaller when <1), which could result in decreased accuracy or even inf. Can you check if it still works gor you when you add a `- multiply: 1000` filter to the base sensor?

**Comment by synco** [comment]
@OttoWinter The reason I made the change was because the filter was causing issues. I would understand it if it was sampling continuously and the numbers changed so fast your eyes could not keep-up, yet here we are sampling for say 200 ms and then waiting 10 seconds (50 sampling blocks later) before next sampling block. A lot can happen in that time so the influence of the previous block is irrelevant, and definitely not more accurate. Easiest way to improve accuracy is to sample for longer.

  If you like to keep the DC component, I suggest we make the block expose the other easy statistical values for a block of samples:
* Mean (DC component)
* Standard deviation (AC component)
* RMS (AC+DC combined)
* Min value
* Max value
Guess we'd need to call it something other than 'ct_clamp'. Other option is to convert the ct_clamp into a block sampler, and then extending the base sensor filters.
eg.
  - platform: ads1115
    id: ct1
    multiplexer: 'A3_GND'
    gain: 4.096
    internal: True
    name: "CT4 Raw HWC"

  - platform: blocksampler
    sensor: ct1
    name: "AC voltage"
    sample_window: 200ms
    update_interval: 10s
    filters:   # one of the following
      - rms
      - mean
      - min
      - max
      - stddev

Is your concern with large floats purely presentation? -because I agree, no point in showing say 2 decimal places for a number in the millions. Yet from a calculation point, I do not see any issue with using floats, unless we are taking 100000's of samples.

**Comment by synco** [comment]
Just got a message mentioning this PR has gone stale... I've been using this on my home system for months and sure beats the undesirable low-pass filter calibration fudge. Hate to see this change go to waste, so  what needs to be done to get it incorporated?

File: esphome/components/ct_clamp/ct_clamp_sensor.cpp [hunk 1]
```diff
@@ -8,17 +8,7 @@ namespace ct_clamp {
...
-  this->high_freq_.start();
-  this->set_timeout("calibrate_offset", this->sample_duration_, [this]() {
-    this->high_freq_.stop();
-    this->is_calibrating_offset_ = false;
-    if (this->num_samples_ != 0) {
-      this->offset_ = this->sample_sum_ / this->num_samples_;
-    }
-  });
-}
+void CTClampSensor::setup() {}
```
Comment (jesserockz) [inline esphome/components/ct_clamp/ct_clamp_sensor.cpp]: Remove the function if it is empty

**Review by jesserockz — APPROVED** [review]

**Pull request merged**